src/test/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketServerClientUnitTest.java

```diff
@@ -0,0 +1,400 @@
+        Response response = mock(Response.class);
+        ArgumentCaptor<Request> captor = ArgumentCaptor.forClass(Request.class);
+
+        when(client.newCall(any())).thenReturn(call);
+        when(call.execute()).thenReturn(response);
+        when(response.isSuccessful()).thenReturn(true);
+
+        when(mapper.writeValueAsString(report)).thenReturn("{payload}");
+
+        // when
+        underTest.uploadReport("project", "repository", "commit", report);
+
+        // then
+        verify(client, times(1)).newCall(captor.capture());
+        Request request = captor.getValue();
+        assertEquals("PUT", request.method());
+        assertEquals("https://my-server.org/rest/insights/1.0/projects/project/repos/repository/commits/commit/reports/com.github.mc1arke.sonarqube", request.url().toString());
+    }
+
+    @Test(expected = BitbucketException.class)
+    public void testUploadReportFails() throws IOException {
+        // given
+        CodeInsightsReport report = mock(CodeInsightsReport.class);
+        Call call = mock(Call.class);
+        Response response = mock(Response.class);
+
+        when(client.newCall(any())).thenReturn(call);
+        when(call.execute()).thenReturn(response);
+        when(response.isSuccessful()).thenReturn(false);
+        when(response.body()).thenReturn(null);
+
+        when(mapper.writeValueAsString(report)).thenReturn("{payload}");
+
+        // when
+        underTest.uploadReport("project", "repository", "commit", report);
+
+        // then
+        fail();
+    }
+
+    @Test()
+    public void testUploadReportFailsWithMessage() throws IOException {
+        // given
+        ErrorResponse.Error error = new ErrorResponse.Error("error!");
+        ErrorResponse errorResponse = new ErrorResponse(Sets.newHashSet(error));
+
+        CodeInsightsReport report = mock(CodeInsightsReport.class);
+        Call call = mock(Call.class);
+        Response response = mock(Response.class);
+        ResponseBody responseBody = mock(ResponseBody.class);
+        ObjectReader reader = mock(ObjectReader.class);
+
+        when(client.newCall(any())).thenReturn(call);
+        when(call.execute()).thenReturn(response);
+        when(response.isSuccessful()).thenReturn(false);
+        when(response.body()).thenReturn(responseBody);
+        when(responseBody.string()).thenReturn("error!");
+        when(response.code()).thenReturn(400);
+
+        when(mapper.writeValueAsString(report)).thenReturn("{payload}");
+
+        when(mapper.reader()).thenReturn(reader);
+        when(reader.forType(ErrorResponse.class)).thenReturn(reader);
+        when(reader.readValue(any(String.class))).thenReturn(errorResponse);
+
+
+        // when
+        try {
+            underTest.uploadReport("project", "repository", "commit", report);
+            fail();
+        } catch (BitbucketException e) {
+            assertTrue(e.isError(400));
+            assertEquals("error!", e.getMessage());
+        }
```

Comment:
Potentially use AssertJ here:

```suggestion
        assertThatThrownBy(() -> underTest.uploadReport("project", "repository", "commit", report))
                .isInstanceOf(BitbucketException.class)
                .hasMessage("error!")
                .extracting(e -> ((BitbucketException) e).isError(400))
                .isEqualTo(true);
```
I'm also not sure what the different between this and the previous test method is, but a similar change could be made there,

Reply:
That looks way cleaner thank you!